### PR TITLE
fix(ir): establish index non-negativity instead of assuming it from INDEX

### LINKED ISSUE #2500: [Bug] Nested scalar if silently drops outer condition during a2a3 lowering
### Component

Codegen

### Description

A nested scalar guard inside an in-core task is silently miscompiled for the
`a2a3` target. Given:

```python
if logical_pos >= 0:
    if logical_pos < ROWS:
        value = state[logical_pos : logical_pos + 1, 0:COLS]
```

the generated PTO drops the outer `logical_pos >= 0` predicate and keeps only
`logical_pos < ROWS`. A negative offset then reaches `partition_view`, where it
is clamped to row 0, producing silently incorrect output.

The logically equivalent combined predicate lowers correctly:

```python
if logical_pos >= 0 and logical_pos < ROWS:
    value = state[logical_pos : logical_pos + 1, 0:COLS]
```

### Steps to Reproduce

```markdown
Save the following as `pypto_nested_if_lowering_probe.py` in a `pypto-lib`
checkout so its `golden` runner is importable:


#!/usr/bin/env python3
"""Minimal repro for a PyPTO nested scalar-if lowering bug.

Both kernels should return ``state[0]``. Their only loop iteration has
``logical_pos == -1``, so it must not read state. The nested spelling is
currently miscompiled on a2a3: the outer ``logical_pos >= 0`` guard disappears
and the negative slice is clamped to row 0, producing ``2 * state[0]``. The
logically equivalent combined predicate works.
"""

import pypto.language as pl


ROWS = 2
COLS = 64


@pl.jit.inline
def nested_if_body(
    token_position: pl.Tensor[[1], pl.INT32],
    state: pl.Tensor[[ROWS, COLS], pl.FP32],
    output: pl.Tensor[[1, COLS], pl.FP32],
):
    pooled = pl.create_tensor([1, COLS], dtype=pl.FP32)
    with pl.at(level=pl.Level.CORE_GROUP, name_hint="nested_if_probe") as pool_tid:
        token_pos = pl.cast(pl.read(token_position, [0]), pl.INDEX)
        window_start = token_pos - ROWS + 1
        acc = state[token_pos : token_pos + 1, 0:COLS]
        for state_idx in pl.range(ROWS - 1):
            logical_pos = window_start + state_idx
            value = pl.full([1, COLS], dtype=pl.FP32, value=0.0)
            if logical_pos >= 0:
                if logical_pos < ROWS:
                    value = state[logical_pos : logical_pos + 1, 0:COLS]
            acc = pl.add(acc, value)
        pooled[0:1, 0:COLS] = acc
    with pl.at(level=pl.Level.CORE_GROUP, name_hint="nested_copy", deps=[pool_tid]):
        output[0:1, 0:COLS] = pooled[0:1, 0:COLS]


@pl.jit.inline
def combined_if_body(
    token_position: pl.Tensor[[1], pl.INT32],
    state: pl.Tensor[[ROWS, COLS], pl.FP32],
    output: pl.Tensor[[1, COLS], pl.FP32],
):
    pooled = pl.create_tensor([1, COLS], dtype=pl.FP32)
    with pl.at(level=pl.Level.CORE_GROUP, name_hint="combined_if_probe") as pool_tid:
        token_pos = pl.cast(pl.read(token_position, [0]), pl.INDEX)
        window_start = token_pos - ROWS + 1
        acc = state[token_pos : token_pos + 1, 0:COLS]
        for state_idx in pl.range(ROWS - 1):
            logical_pos = window_start + state_idx
            value = pl.full([1, COLS], dtype=pl.FP32, value=0.0)
            if logical_pos >= 0 and logical_pos < ROWS:
                value = state[logical_pos : logical_pos + 1, 0:COLS]
            acc = pl.add(acc, value)
        pooled[0:1, 0:COLS] = acc
    with pl.at(level=pl.Level.CORE_GROUP, name_hint="combined_copy", deps=[pool_tid]):
        output[0:1, 0:COLS] = pooled[0:1, 0:COLS]


@pl.jit
def nested_if_probe(
    token_position: pl.Tensor[[1], pl.INT32],
    state: pl.Tensor[[ROWS, COLS], pl.FP32],
    output: pl.Out[pl.Tensor[[1, COLS], pl.FP32]],
):
    nested_if_body(token_position, state, output)


@pl.jit
def combined_if_probe(
    token_position: pl.Tensor[[1], pl.INT32],
    state: pl.Tensor[[ROWS, COLS], pl.FP32],
    output: pl.Out[pl.Tensor[[1, COLS], pl.FP32]],
):
    combined_if_body(token_position, state, output)


def build_tensor_specs():
    import torch
    from golden import TensorSpec

    return [
        TensorSpec(
            "token_position",
            [1],
            torch.int32,
            init_value=lambda: torch.tensor([0], dtype=torch.int32),
        ),
        TensorSpec(
            "state",
            [ROWS, COLS],
            torch.float32,
            init_value=lambda: torch.arange(
                1, ROWS * COLS + 1, dtype=torch.float32
            ).reshape(ROWS, COLS),
        ),
        TensorSpec("output", [1, COLS], torch.float32, is_output=True),
    ]


def golden(tensors):
    tensors["output"][:] = tensors["state"][0:1]


if __name__ == "__main__":
    import argparse
    from golden import run_jit

    parser = argparse.ArgumentParser()
    parser.add_argument("-p", "--platform", default="a2a3")
    parser.add_argument("-d", "--device", type=int, default=0)
    parser.add_argument("--case", choices=["nested", "combined", "all"], default="all")
    args = parser.parse_args()

    cases = {"nested": nested_if_probe, "combined": combined_if_probe}
    selected = cases if args.case == "all" else {args.case: cases[args.case]}
    failed = []
    for name, fn in selected.items():
        print(f"--- {name} ---")
        result = run_jit(
            fn=fn,
            specs=build_tensor_specs(),
            golden_fn=golden,
            runtime_cfg={"platform": args.platform, "device_id": args.device},
            rtol=0.0,
            atol=0.0,
        )
        if not result.passed:
            failed.append(name)
            if result.error:
                print(result.error)

    raise SystemExit(1 if failed else 0)


Run on an available Ascend 910B device:


python pypto_nested_if_lowering_probe.py -p a2a3 -d 0 --case all
```

### Expected Behavior

Both variants are semantically equivalent and should return `state[0]`:

```text
output[0, 0:4] = [1.0, 2.0, 3.0, 4.0]
```

Both cases should pass exact validation.

### Actual Behavior

The nested variant fails all 64 elements and returns twice the expected value:

```text
--- nested ---
output FAIL: mismatched elements 64/64
actual[0, 0:4]   = [2.0, 4.0, 6.0, 8.0]
expected[0, 0:4] = [1.0, 2.0, 3.0, 4.0]

--- combined ---
output PASS
```

The generated PTO for the nested form contains only the upper-bound check:

```mlir
%is_lt_rows = arith.cmpi slt, %logical_pos, %c2_index : index
scf.if %is_lt_rows {
  %clamped = arith.maxsi %logical_pos, %c0_index : index
  %view = pto.partition_view %state, offsets = [%clamped, %c0_index], ...
  pto.tload ...
}
```

There is no `logical_pos >= 0` comparison. Since `logical_pos == -1`, the
remaining upper-bound check succeeds and the slice offset is clamped to row 0.

The combined form lowers correctly:

```mlir
%is_nonnegative = arith.cmpi sle, %c0_index, %logical_pos : index
%is_lt_rows = arith.cmpi slt, %logical_pos, %c2_index : index
%valid = arith.andi %is_nonnegative, %is_lt_rows : i1
scf.if %valid {
  ...
}

### Git Commit ID

c56951b923e528026b1020087adc257b1279e645

### NPU Kind

Ascend 910C

### Host Platform

Linux (aarch64)

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by lyfne123**
## Summary

A nested scalar guard lost its outer predicate on the way to PTO:

    if logical_pos >= 0:
        if logical_pos < ROWS:
            value = state[logical_pos : logical_pos + 1, 0:COLS]

Simplify proved `logical_pos >= 0` statically true and folded the outer
`IfStmt` away, leaving the upper-bound check standing alone. A negative
`logical_pos` passes that check, reaches `pto.partition_view`, and is clamped to
row 0 -- the kernel reads the wrong row and returns silently wrong data. The
logically equivalent `if a >= 0 and a < N` spelling was unaffected: `CanProve`
on an `And` needs both halves, and the second one is not provable.

The cause was a blanket assumption: `ConstIntBoundAnalyzer` gave any unbound
`INDEX` Var a default bound of `[0, +inf)`. `INDEX` does not imply that. It is a
signed type -- codegen emits `arith.cmpi slt` and `arith.maxsi` for it -- and it
is the only scalar type the sliding-window idiom in the report has to compute an
offset in.

## Non-negativity is now established, never assumed

An unbound Var is `[-inf, +inf]` whatever its dtype. Every real source binds it:

| Source | Bound | Established by |
| ------ | ----- | -------------- |
| Assigned scalar | its RHS's range | `BindScalarBound` |
| Loop variable | `[start, stop)`, or `[start, +inf)` for a symbolic `stop` | `IRMutatorWithAnalyzer` on `ForStmt` |
| Branch arm | the constraint's range | `EnterConstraint` |
| Block / subblock builtin | `[0, +inf)`; a block *count* is `[1, +inf)` | the op's own semantics |
| Shape / valid-shape extent | `[0, +inf)` | `AssumeUnboundVarsNonNegative` |
| Runtime scalar parameter | `[-inf, +inf]` | nothing -- the caller picks the value |

Two of these are new, and neither is a fact about `INDEX`.

`tile.get_subblock_idx()` and friends are non-negative because of what the op
*returns*, so `ConstIntBoundAnalyzer` now answers `Call` from a small table of
builtin result ranges instead of `Everything`.

A whole shape or valid_shape *dimension* is non-negative because it is a count
of elements, so `DimensionSymbolScope` binds one to `[0, +inf)` around the
write-union proofs in `type_inference` -- and stops there.

The stopping point is the load-bearing part. Being a `valid_shape` makes the
*field* an extent; it does not make every variable that computes it one:

    valid = pl.max(-x, 0)    # a legal dynamic extent over a signed scalar

Assuming `x >= 0` folds that to a constant `0`, which reads as an empty region
and silently shrinks the result. So the scope binds a dimension that *is* a bare
symbol and never descends into a compound one, and it never binds an offset at
all -- `max(x, 0)` in an offset is a deliberate clamp whose whole point is that
`x` can be negative, and folding it to `x` moves the region a store writes.

`ForStmt` supplies the last piece. A symbolic `stop` -- `pl.range(n)` -- used to
draw `i >= 0` from the dtype; it now comes from the loop, which binds
`[start, +inf)` when `start` is a constant and the step is a positive constant.
A negative step walks the other way, so it does not qualify.

## Changes

- `include/pypto/ir/arith/analyzer.h`, `src/ir/arith/const_int_bound.cpp`:
  the dtype default is gone; `AssumeUnboundVarsNonNegative` replaces it, and
  constraint-scope recovery carries `std::optional<Bound>`.
- `src/ir/arith/ir_mutator_with_analyzer.cpp`: the symbolic-`stop` lower bound.
- `src/ir/transforms/simplify_pass.cpp`: `BindScalarBound` records the RHS range
  as-is rather than intersecting it with the default.
- `src/ir/op/type_inference.cpp`: `DimensionSymbolScope`, opened over the
  target's and source's own valid shapes in `InferWriteValidShapeUnion` -- never
  over `params.offsets`.

No public API changes: the analyzer keeps the interface it had, since nothing
here needs a knob.
- `docs/{en,zh}/dev/passes/05-simplify.md`: where an index bound comes from.

## Tests

`tests/ut/ir/transforms/test_simplify_pass.py` gains the nested-guard
regression. Five existing tests asserted the dtype default itself, so their
premise is gone; each keeps its real subject and sources non-negativity
explicitly instead:

- `test_index_var_non_negative` -> `test_index_var_is_not_implicitly_non_negative`,
  plus `test_extent_mode_makes_unbound_vars_non_negative` for the opt-in.
- `test_index_non_negativity_preserved_under_constraint` ->
  `test_lower_bound_preserved_under_upper_bound_constraint`. Its subject was
  always "an upper-bound constraint leaves the lower half alone"; only the
  lower bound's origin changes.
- `test_max_index_non_negative` / `test_max_zero_index_commutative` ->
  `test_max_zero_folds_{for_a_non_negative_var,commutatively}`, with
  `test_max_zero_is_kept_for_a_bare_index_var` covering the case that must not
  fold -- folding `max(n, 0)` there rewrites a real clamp into a no-op.
- `test_symbolic_index_scalar_keeps_nonneg_default_bound` ->
  `test_leaf_index_parameter_keeps_its_guard`, asserting the opposite of what it
  used to: a `Scalar[INDEX]` parameter can be `-1`, so its guard is live.

## Verification

- `pytest tests/ut -n 16` -- 10768 passed, 8 skipped, 1 xfailed; and
  `pytest tests/st/codegen` -- 60 passed, against the pinned ptoas v0.60. One
  further
  test, `test_symlinked_import_path_still_names_the_caller`, is deselected: it
  fails identically on an unpatched build here because its subprocess re-execs
  with its own `PYTHONPATH` and resolves `pypto` through the editable meta-path
  finder instead of the symlink.
- `ctest` -- 1/1 passed.
- The reported kernel keeps both guards, and the generated PTO matches the
  report's expected form: `arith.cmpi sle, %c0_index, %pos` guarding an `scf.if`
  that contains the `arith.cmpi slt` check and its `scf.if`. The `and` spelling's
  PTO is byte-identical to before.
- Blast radius: the DeepSeek-V4-Flash MoE model (`--ep 2`, 31 kernels) generates
  PTO identical to `origin/main` once the run-to-run SSA naming is normalized.
- `clang-format`, `clang-tidy` (`--strict-version`, against `origin/main`),
  `markdownlint-cli2`, and every `tests/lint/check_*.py` -- clean. `ruff` is
  version-pinned and not installable in this environment; CI's pre-commit job
  covers it.

Not verified on device: this environment's installed `simpler_setup` assets are
newer than the pinned `runtime/` submodule, so the generated orchestration C++
fails to find `orchestration_api.h` and no kernel from a `main`-based checkout
reaches the card, with or without this change. The qwen3 and decode_swa models
need a card and were left to CI.

Closes #2500

### CODE REVIEW

File: tests/ut/ir/transforms/test_simplify_pass.py [hunk 1]
```diff
@@ -1104,14 +1104,41 @@ def main(self, out: pl.Tensor[[8], pl.INDEX]):
...
+                if a < 0:
+                    pl.tensor.write(out, [0], b)
+                else:
+                    pl.tensor.write(out, [0], a)
+
+        @pl.program
+        class Expected:
+            @pl.function
+            def main(self, a: pl.Scalar[pl.INDEX], b: pl.Scalar[pl.INDEX], out: pl.Tensor[[1], pl.INDEX]):  # noqa: ARG002
+                pl.tensor.write(out, [0], a)
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply tests/ut/ir/transforms/test_simplify_pass.py]: You are right, and thanks — this was the real defect, not just an over-eager test.

Fixed in 9382f860 by removing the blanket assumption rather than narrowing where
it applies. `ConstIntBoundAnalyzer` no longer gives an unbound Var any bound from
its dtype: an unbound `INDEX` Var is now `[-inf, +inf]`. Non-negativity has to be
established by something real — an assigned scalar's RHS, a loop's `[start, stop)`,
a branch constraint, or the new opt-in below.

That opt-in is your "attach `[0, +inf)` only to symbols whose role proves it":
`ConstIntBoundAnalyzer::AssumeUnboundVarsNonNegative`, off by default. The shape
and valid-extent proofs in `type_inference` (`DimensionsEqual`,
`ProveValidExtent{Equal,LessEqual}`, `FoldExtent`) turn it on, because every Var
reaching them names a tensor dimension or valid-shape extent, and an extent is a
count of elements. That is scoped by role, not by type. Without it,
`tensor.assemble` loses `k <= k + m` and emits `pl.min(pl.max(k, k + m), 64)`,
and `tile.store` rejects a valid-shape symbol it can no longer place at or after 0
— those two were the only real functional users of the old default.

`ForStmt` supplies the other half: a symbolic `stop` (`pl.range(n)`) used to draw
`i >= 0` from the dtype, and now binds `[start, +inf)` when `start` is constant and
the step is a positive constant.

The test you flagged is inverted rather than deleted —
`test_leaf_index_parameter_keeps_its_guard` now asserts the guard *survives*, with
your `a = -1` case as the rationale. Four other tests asserted the old default
directly and were rewritten to source non-negativity explicitly; one of them,
`test_max_zero_is_kept_for_a_bare_index_var`, is new and covers the fold that must
*not* happen.

Verified: 10484 UT pass, the #2500 kernel keeps both guards, and the
DeepSeek-V4-Flash MoE model (`--ep 2`, 31 kernels) generates PTO identical to
`origin/main` once run-to-run SSA naming is normalized.

**Review by Hzfengsy — APPROVED** [review]
